OrbitCore/LinuxEventTracer.cpp

```diff
@@ -0,0 +1,262 @@
+//-----------------------------------
+// Copyright Pierric Gimmig 2013-2019
+//-----------------------------------
+
+#include "ContextSwitch.h"
+#include "OrbitProcess.h"
+#include "PrintVar.h"
+#include "LinuxPerfUtils.h"
+#include "LinuxPerfEvent.h"
+#include "LinuxEventTracer.h"
+#include "LinuxUtils.h"
+#include "Utils.h"
+#include "Capture.h"
+#include "TimerManager.h"
+
+#include <assert.h>
+#include <functional>
+#include <queue>
+#include <vector>
+
+
+//-----------------------------------------------------------------------------
+LinuxEventTracer::LinuxEventTracer()
+{
+    //m_Callback = a_Callback ? a_Callback : [this](const std::string& a_Buffer)
+    //{
+    //    CommandCallback(a_Buffer);
+    //};
+}
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Start()
+{
+    PRINT_FUNC;
+#if __linux__
+    m_ExitRequested = false;
+
+    m_Thread = std::make_shared<std::thread>(&LinuxEventTracer::foo, &m_ExitRequested);
+
+   /* m_Thread = std::make_shared<std::thread>
+        ( &LinuxUtils::StreamCommandOutput
+        , "perf record -e sched:sched_switch -o - | perf script -i -"
+        , m_Callback
+        , &m_ExitRequested );
+    */
+    m_Thread->detach();
+#endif
+}
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::Stop()
+{
+    m_ExitRequested = true;
+}
+
+
+//-----------------------------------------------------------------------------
+void LinuxEventTracer::CommandCallback(std::string a_Line)
+{
+    std::vector<std::string> tokens = Tokenize(a_Line);
+    
+    // TODO: log broken lines
+    if (tokens[4] != "sched:sched_switch:" || tokens[8] != "==>")
+    {
+        return;
+    }
+
+    auto oldTID = atoi(tokens[1].c_str());
+    auto cpu = atoi(trim(tokens[2], "[]").c_str());
+    auto time = GetMicros(tokens[3])*1000;
+    char state = tokens[7].at(0);
+
+    auto newCommCPUTID = Tokenize(tokens[9], "/:");
+    auto newTID = atoi(newCommCPUTID.back().c_str());
+
+    // TODO: only process context switches that relate to the traced process.
+
+    ContextSwitch::SwitchType type = ContextSwitch::Invalid;
+    int tid = -1;
+    switch (state)
+    {
+    // sleep:
+    case 'S':
+        //if (!Capture::GTargetProcess->HasThread(oldTID))
+        //    return;
+
+        type = ContextSwitch::Out;
+        tid = oldTID;
+        break;
+    
+    // run:
+    case 'R':
+        //if (!Capture::GTargetProcess->HasThread(newTID))
+        //    return;
+
+        type = ContextSwitch::In;
+        tid = newTID;
+        break;
+
+    // TODO: handle the other cases
+
+    default:
+        return;
+    }
+
+    if (state != ContextSwitch::Invalid && tid != -1)
+    {
+        ++Capture::GNumContextSwitches;
+
+        ContextSwitch CS ( type );
+        CS.m_ThreadId = tid;
+        CS.m_Time = time;
+        CS.m_ProcessorIndex = cpu;
+        //TODO: Is this correct?
+        CS.m_ProcessorNumber = cpu;
```

Comment:
What's the meaning of ProcessorIndex vs. ProcessorNumber? Does it make sense to track them individually? (Are there places where they are actually different?)

Reply:
ProcessorNumber is the number of CPUs in the machine, ProcessorIndex is the core index. Currently ProcessorNumber is not used by Orbit at all. In particular, I don't think that this information is available while tracing. We could use 0 for ProcessorNumber otherwise. In the end, it basically needs to be consistent.